game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -668,92 +669,71 @@ private void removeAirNoLongerInTerritory() {
-        final Collection<Unit> transports =
-            CollectionUtils.getMatches(attackingUnits, Matches.unitCanTransport());
-        for (final Unit unit : transports) {
-          change.add(ChangeFactory.unitPropertyChange(unit, true, Unit.WAS_IN_COMBAT));
-        }
-        bridge.addChange(change);
+    // If any attacking transports are in the battle, set their status to later restrict
+    // load/unload
+    final Collection<Unit> transports =
+        CollectionUtils.getMatches(
+            attackingUnits, Matches.unitCanTransport().and(Matches.unitIsOwnedBy(attacker)));
+    if (!transports.isEmpty()) {
+      final CompositeChange change = new CompositeChange();
+      for (final Unit unit : transports) {
+        change.add(ChangeFactory.unitPropertyChange(unit, true, Unit.WAS_IN_COMBAT));
       }
+      bridge.addChange(change);
     }
-    // write attacking units to history
-    if (!attackingUnits.isEmpty()) {
-      bridge.getHistoryWriter().addChildToEvent(transcriptText.toString(), allAttackingUnits);
+  }
+
+  private void writeUnitsToHistory(final IDelegateBridge bridge) {
+    if (headless) {
+      return;
     }
-    // find all defending players (unsorted)
-    final Collection<GamePlayer> defenders = new ArrayList<>();
-    for (final GamePlayer current : playerSet) {
-      if (gameData.getRelationshipTracker().isAllied(defender, current)
-          || current.equals(defender)) {
-        defenders.add(current);
+    final Set<GamePlayer> playersWithUnits = battleSite.getUnitCollection().getPlayersWithUnits();
+
+    final Collection<GamePlayer> attackers = findAllies(playersWithUnits, attacker);
+    addPlayerCombatHistoryText(attackers, attackingUnits, true, bridge.getHistoryWriter());
+    final Collection<GamePlayer> defenders = findAllies(playersWithUnits, defender);
+    addPlayerCombatHistoryText(defenders, defendingUnits, false, bridge.getHistoryWriter());
+  }
+
+  private static Collection<GamePlayer> findAllies(
+      final Collection<GamePlayer> candidatePlayers, final GamePlayer player) {
+    final var relationshipTracker = player.getData().getRelationshipTracker();
+    final Collection<GamePlayer> allies = new ArrayList<>();
+    for (final GamePlayer current : candidatePlayers) {
+      if (current.equals(player) || relationshipTracker.isAllied(player, current)) {
+        allies.add(current);
       }
     }
-    final StringBuilder transcriptBuilder = new StringBuilder();
-    // find all defending units (unsorted)
-    final Collection<Unit> allDefendingUnits = new ArrayList<>();
-    for (final Iterator<GamePlayer> defendersIter = defenders.iterator();
-        defendersIter.hasNext(); ) {
-      final GamePlayer current = defendersIter.next();
-      final String delim;
-      if (defendersIter.hasNext()) {
-        delim = "; ";
-      } else {
-        delim = "";
+    return allies;
+  }
+
+  private void addPlayerCombatHistoryText(
```

Comment:
This is OK because the following two existing findings are fixed:

```
Method writeUnitsToHistory has a Cognitive Complexity of 26 (exceeds 7 allowed). Consider refactoring.
Method writeUnitsToHistory has 78 lines of code (exceeds 30 allowed). Consider refactoring.
```

https://codeclimate.com/github/triplea-game/triplea/pull/6529